### PR TITLE
contributing legit

### PR DESCRIPTION
**Pull request by ericmjl**

### CODE REVIEW

File: recipes/legit/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+{% set packagename = 'legit' %}
+{% set version = "0.2.0" %}
+{% set md5 = "5df3cd402b0de8f0251a87904b5eb88c" %}
+
+package:
+  name: {{ packagename }}
+  version: {{ version }}
+
+source:
+  fn: {{ packagename }}-{{ version }}.tar.gz
```
Comment (ericdill) [inline recipes/legit/meta.yaml]: change your filename to end with ".zip" since you're downloading a zipball from pypi, not a tarball

File: recipes/legit/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  entry_points:
+    - legit = legit.cli:main
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - clint >=0.4.1
+    - gitpython >=0.3.4
+    - six >=1.9.0
```
Comment (ericdill) [inline recipes/legit/meta.yaml]: Because you're doing `--single-version-externally-managed`, you can drop clint, gitpython and six as build requirements as this ignores any requirements specified in setup.py. They should remain as run requirements

File: recipes/legit/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+test:
+  # Some package might need a `test/commands` key to check CLI.
+  imports:
+    - {{ packagename }}
+
+  commands:
+    - legit --help
```
Comment (ericdill) [inline recipes/legit/meta.yaml]: The `legit` cli needs a git repository to work off of or you just get this message:

```
$ legit
Not a git repository.
```
and the real kicker is that it [exits 128] (https://github.com/kennethreitz/legit/blob/v0.2.0/legit/scm.py#L47), which means that your test section will always fail if you dont have a git repository.

Another option would be to change the source section to be:

```yaml
source:
  git_url: https://github.com/kennethreitz/legit
  git_tag: v{{ version }}
```

and then change the commands section to be:

```yaml
commands:
  - cd {{ environ.SRC_DIR }} && legit --help
```

But *that* [exits 1] (https://github.com/kennethreitz/legit/blob/v0.2.0/legit/cli.py#L53) after printing the help message 

So do the [`-v` or `--version` flags] (https://github.com/kennethreitz/legit/blob/v0.2.0/legit/cli.py#L57).

I guess the best option is to just to use `legit help`, which exercises the CLI and prints out:

```
>legit help
help <command>

Display help for legit command.
```

and, most importantly, exits with a 0, so it does not cause conda build to fail!


-------

@jakirkham this seems like a reasonable exception to the "don't clone the git repo" rule, right?

tl; dr

source should look like:
```yaml
source:
  git_url: https://github.com/kennethreitz/legit
  git_tag: v{{ version }}
```

commands should look like:

```yaml
commands:
  - cd {{ environ.SRC_DIR }} && legit --help
```

Comment (ericmjl) [inline recipes/legit/meta.yaml]: For now, I will use `legit --version`.

**Comment by ericdill** [comment]
@ericmjl I'm sorry.  Reading back over my comment, I see that I was not communicating clearly.  `legit --version` returns `128` which is going to be interpreted as an error code by conda.  Using any `legit` command other than `-v`, `--version`, `-h`, `--help` will return `0` and result in conda-build passing the test section.  For example, `legit help` will work in the test section.  Sorry again for not communicating clearly

**Comment by ericmjl** [comment]
@ericdill no worries, I appreciate the help in getting the recipe going! :smile: I'm going to push up another change, let's see if it'll work.

**Comment by ericdill** [comment]
Well yay, at least the tests are passing on os-x and linux.  That solved some of the problem

It is unclear to me why appveyor is failing on not being able to find six.moves.  I've asked for help in the gitter channel. Hopefully someone with more windows/appveyor experience can lend a hand here as I cannot replicate that `six.moves` import error locally on my windows box.  (maybe @ocefpaf ? I see you are at least a little bit to (git) blame for the `appveyor.yml` configuration)

**Comment by msarahan** [comment]
Sorry, I don't have any immediate advice here.

```
  File "C:\Miniconda\conda-bld\work\legit.py", line 3, in <module>
    from legit.cli import main
  File "C:\Miniconda\conda-bld\work\legit\__init__.py", line 3, in <module>
    from .core import *
  File "C:\Miniconda\conda-bld\work\legit\core.py", line 10, in <module>
    from . import bootstrap
  File "C:\Miniconda\conda-bld\work\legit\bootstrap.py", line 11, in <module>
```

This looks suspect.  I don't know if those paths are actually relevant, or relics of the pyc process.  Those are suspect for two reasons:

1) They are not the test prefix
2) They are also not the build prefix - they are the source prefix.

File: recipes/legit/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  # Some package might need a `test/commands` key to check CLI.
+  imports:
+    - {{ packagename }}
+
+  commands:
+    - cd {{ environ.SRC_DIR }} && legit help
+
+about:
+  home: https://github.com/kennethreitz/legit
+  license: BSD
```
Comment (msarahan) [inline recipes/legit/meta.yaml]: How many clauses?

File: recipes/legit/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    - gitpython >=0.3.4
+    - six >=1.9.0
+
+test:
+  # Some package might need a `test/commands` key to check CLI.
+  imports:
+    - {{ packagename }}
+
+  commands:
+    - cd {{ environ.SRC_DIR }} && legit help
```
Comment (msarahan) [inline recipes/legit/meta.yaml]: Oh, I see it now.  environ.SRC_DIR is totally not guaranteed to exist at test time.  You should instead clone a simple repo as part of your test.  You may want to turn to creating run_test.bat/run_test.sh, or run_test.py if you know how to clone/create a repo from python.  Feel free to use legit itself if that helps.

Comment (msarahan) [inline recipes/legit/meta.yaml]: On second thought, that's silly.  Your tests wouldn't be running on the other platforms.  I still think this assumption is unsafe in general, but I don't see why it isn't working here.

Comment (msarahan) [inline recipes/legit/meta.yaml]: OK, now I think I get it.  This is tricky.  I believe it is caused by your cd into the source folder.  In doing so, the legit entry point finds the local legit folder there, rather than the installed one.  Things spiral from there.  I have no good answer for why this does not disrupt the other platforms.  I have confounded myself in this situation before several times, locally, though, and this seems familiar.

In light of this, I still recommend a separate git repo here for testing - cloned or created via gitpython, or perhaps use the conda git package as a test requirement if you need to.  Avoid a cd into the source folder.

Comment (jakirkham) [inline recipes/legit/meta.yaml]: > I still recommend a separate git repo here for testing

Nothing more to add. 😄

File: recipes/xray-vision/meta.yaml [hunk 6]
```diff
@@ -1,50 +0,0 @@
-{% set version = "0.0.6" %}
```
Comment (jakirkham) [inline recipes/xray-vision/meta.yaml]: Should be cleared now. So please try rebasing to get rid of this.

File: recipes/legit/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/legit/meta.yaml]: So I agree with @msarahan we should not be assuming that `SRC_DIR` exists in tests. It might not be the case. Even if it does exist it, it can confound problems as it did here. His advice to clone some other repo for testing seems sound. Please also add `git` as a test dependency.

File: recipes/legit/meta.yaml (hunk 5 above)
Comment (ericdill) [inline recipes/legit/meta.yaml]: Thanks for the help here @msarahan . 

@ericmjl try this:
```diff
diff --git a/recipes/legit/meta.yaml b/recipes/legit/meta.yaml
index 4aa0ab0..77bc633 100644
--- a/recipes/legit/meta.yaml
+++ b/recipes/legit/meta.yaml
@@ -32,7 +32,15 @@ test:
     - {{ packagename }}
 
   commands:
-    - cd {{ environ.SRC_DIR }} && legit help
+    - mkdir test-git-repo
+    - cd test-git-repo
+    - git init
+    # legit needs a remote or it exits with a non-zero status
+    # so we just need to add a fake one
+    - git remote add origin http://
+    - legit help
+  requires:
+    - git
 
 about: 
   home: https://github.com/kennethreitz/legit
```

It passes locally on my linux laptop. I cannot guarantee that this works on windows.  If it doesn't then you'll need to put the content of `commands` into a `build.sh` file and then translate the `build.sh` into a `build.bat` to make windows happy.

File: recipes/legit/meta.yaml (hunk 4 above)
Comment (ericmjl) [inline recipes/legit/meta.yaml]: Hmm... I think it's 3-clause. Is that right?

https://github.com/kennethreitz/legit/blob/develop/LICENSE

File: recipes/legit/meta.yaml (hunk 5 above)
Comment (ericmjl) [inline recipes/legit/meta.yaml]: Okay, let me give that a shot.

File: recipes/legit/meta.yaml (hunk 4 above)
Comment (ericdill) [inline recipes/legit/meta.yaml]: Yeah that's 3-clause

**Comment by ericmjl** [comment]
alrighty, I think the recipe works - all checks have passed!

**Comment by msarahan** [comment]
Nice work @ericmjl @ericdill @jakirkham - merging.

**Comment by ericmjl** [comment]
thanks @msarahan! learning each time you guys help 😄

**Comment by jakirkham** [comment]
Awesome great job guys.

File: recipes/legit/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+{% set version = "0.2.0" %}
+{% set md5 = "5df3cd402b0de8f0251a87904b5eb88c" %}
+
+package:
+  name: {{ packagename }}
+  version: {{ version }}
+
+source:
+  git_url: https://github.com/kennethreitz/legit
+  git_tag: v{{ version }}
```
Comment (janschulz) [inline recipes/legit/meta.yaml]: Git tags can be "updated", this should be a commit ID or a release.
